core/src/main/java/org/apache/calcite/rel/metadata/JaninoRelMetadataProvider.java

```diff
@@ -97,6 +97,8 @@
   private static final Set<Class<? extends RelNode>> ALL_RELS =
       new CopyOnWriteArraySet<>();
 
+  private static final Set<String> HANDLERS_GENERATED = new HashSet<>();
```

Comment:
For global hashset, would it be better to use a thread-safe version? like ConcurrentHashMap or wrap it with Collections.synchorized(..)?

Reply:
The `JaninoRelMetadataProvider` would in total set to a thread local.

Reply:
Correct me if I am wrong, threads see their own JaninoRelMetadataProvider instance, different threads may touch the static shared HashSet concurrently which would be not safe.

Reply:
Yes, a private local member here is enough.

Reply:
Thanks for the discussion and thanks for the kind reminder.
I have revised the code to wrap the manipulation to the collection in a synchronized block.